File: libusbcamera/src/main/java/com/rain/uvc/utils/CameraNativeUtils.java
package com.rain.uvc.utils;

import android.text.TextUtils;
import android.view.Surface;

import com.rain.uvc.listener.IFrameListener;
import com.rain.uvc.mode.CameraSize;
import com.rain.uvc.state.CameraPreviewFormat;
import com.rain.uvc.state.CameraParameter;
import com.rain.uvc.state.CameraParameterType;
import com.rain.uvc.state.CameraSupportParameters;
import com.rain.uvc.mode.IntRange;
import com.rain.uvc.state.DisplayTransformState;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

/**
 * 调用的native方法
 */
public class CameraNativeUtils {

    static {
        System.loadLibrary("usb100");
        System.loadLibrary("uvc");
        System.loadLibrary("uvcCamera");
    }

    public static native boolean debuggable(int status);

    /**
     * 根据对应的FileDescriptor连接指定设备
     *
     * @param fd 文件描述符
     * @return 是否成功
     */
    public static native long nativeOpen(int fd, int busNum, int devAddress);

    /**
     * 读取可使用的v4l2的列表信息
     *
     * @return 返回列表地址
     */
    public static native String[] nativeLoadV4L2Devices();

    /**
     * 根据对应的FileDescriptor连接指定设备
     *
     * @param videoPath 对应打开的路径
     * @return 是否成功
     */
    public static native long nativeOpenVideo(String videoPath);

    /**
     * 断开连接设备
     *
     * @param nativeId 对应的内存地址值
     * @return 是否成功
     */
    public static native boolean nativeClose(long nativeId);

    /**
     * 开启预览
     *
     * @param nativeId 对应的内存地址值
     * @return 是否成功
     */
    public static native boolean nativeStartPreview(long nativeId);

    /**
     * 关闭预览
     *
     * @param nativeId 对应的内存地址值
     * @return 是否成功
     */
    public static native boolean nativeStopPreview(long nativeId);

    /**
     * 设置预览控件
     *
     * @param nativeId 对应的内存地址值
     * @param surface  对应预览的surface
     * @return 是否成功
     */
    public static native boolean nativeSetDisplaySurface(long nativeId, Surface surface);

    /**
     * 设置预览的分辨率信息
     *
     * @param nativeId 对应的内存地址值
     * @param width    宽
     * @param height   高
     * @return 是否成功
     */
    public static native boolean nativeSetPreviewSize(long nativeId, int width, int height, int format);

    /**
     * 设置预览监听，返回的数据格式固定为RGB格式
     *
     * @param nativeId 设置的对应id
     * @param listener 监听器
     */
    public static native boolean setPreviewListener(long nativeId, IFrameListener listener, int mode);

    /**
     * 获取支持的参数信息
     */
    public static native String nativeGetSupportedParameters(long nativeId, int type);


    /**
     * 获取当前参数信息
     */
    public static native String nativeGetParameterValue(long nativeId, int type);

    /**
     * 设置当前参数信息
     */
    public static native boolean nativeSetParameterValue(long nativeId, int type, int value);

    /**
     * 获取对应的参数信息
     *
     * @param nativeId 对应的内存地址值
     * @param key      对应支持的参数key
     * @return 返回值，如果不支持或者不存在，则返回null
     */
    public static <T> T getParameter(long nativeId, CameraParameter.Key<T> key) {
        if (nativeId == 0L) return null;
        int id = ketTypeToNativeId(key.type);
        if (id == -1) {
            return null;
        }
        String value = nativeGetParameterValue(nativeId, id);
        if (CameraSize.class.isAssignableFrom(key.mClass)) {
            String[] array = null;
            if (!TextUtils.isEmpty(value)) {
                array = value.split(":");
            }
            if (array == null || array.length < 3) {
                return null;
            }
            return (T) new CameraSize(Integer.parseInt(array[0]), Integer.parseInt(array[1]), CameraPreviewFormat.valueToFormatMode(Integer.parseInt(array[2])));
        }
        if (DisplayTransformState.class.isAssignableFrom(key.mClass)) {
            int orientation = -1;
            if (!TextUtils.isEmpty(value)) {
                orientation = Integer.parseInt(value);
            }
            return (T) DisplayTransformState.orientationToState(orientation);
        }

        if (Integer.class.isAssignableFrom(key.mClass)) {
            if (!TextUtils.isEmpty(value)) {
                return (T) Integer.getInteger(value);
            }
            return null;
        }
        if (boolean.class.isAssignableFrom(key.mClass) || Boolean.class.isAssignableFrom(key.mClass)) {
            int state = 0;
            if (!TextUtils.isEmpty(value)) {
                state = Integer.parseInt(value);
            }
            return (T) Boolean.valueOf(state == 1);
        }
        return null;
    }

    /**
     * 设置对应的参数信息
     *
     * @param nativeId 对应的内存地址值
     * @param key      对应支持的参数key
     * @param value    对应的值
     * @return 是否支持
     */
    public static <T> boolean setParameter(long nativeId, CameraParameter.Key<T> key, T value) {
        if (nativeId == 0L) return false;
        if (int.class.isAssignableFrom(key.mClass) || Integer.class.isAssignableFrom(key.mClass)) {
            return nativeSetParameterValue(nativeId, ketTypeToNativeId(key.type), (int) value);
        }
        if (boolean.class.isAssignableFrom(key.mClass) || Boolean.class.isAssignableFrom(key.mClass)) {
            return nativeSetParameterValue(nativeId, ketTypeToNativeId(key.type), (boolean) value ? 1 : 0);
        }
        if (DisplayTransformState.class.isAssignableFrom(key.mClass)) {
            return nativeSetParameterValue(nativeId, ketTypeToNativeId(key.type), ((DisplayTransformState) value).getAngle());
        }
        if (CameraSize.class.isAssignableFrom(key.mClass)) {
            CameraSize size = (CameraSize) value;
            return nativeSetPreviewSize(nativeId, size.getWidth(), size.getHeight(), size.getFormat().getValue());
        }
        return false;
    }

    private static int ketTypeToNativeId(String type) {
        if (CameraParameterType.PREVIEW_SIZE.equals(type)) {
            return 0;
        } else if (CameraParameterType.DISPLAY_TRANSFORM.equals(type)) {
            return 1;
        } else if (CameraParameterType.AUTO_EXPOSURE.equals(type)) {
            return 2;
        } else if (CameraParameterType.EXPOSURE.equals(type)) {
            return 3;
        } else if (CameraParameterType.BRIGHTNESS.equals(type)) {
            return 4;
        } else if (CameraParameterType.CONTRAST.equals(type)) {
            return 5;
        } else if (CameraParameterType.GAIN.equals(type)) {
            return 6;
        } else if (CameraParameterType.SATURATION.equals(type)) {
            return 7;
        } else if (CameraParameterType.ZOOM.equals(type)) {
            return 8;
        } else if (CameraParameterType.AUTO_FOCUS.equals(type)) {
            return 9;
        } else if (CameraParameterType.FOCUS.equals(type)) {
            return 10;
        } else if (CameraParameterType.IRIS.equals(type)) {
            return 11;
        } else if (CameraParameterType.AUTO_HUE.equals(type)) {
            return 12;
        } else if (CameraParameterType.HUE.equals(type)) {
            return 13;
        } else if (CameraParameterType.AUTO_WHITE_BALANCE.equals(type)) {
            return 14;
        } else if (CameraParameterType.WHITE_BALANCE.equals(type)) {
            return 15;
        } else if (CameraParameterType.SCENE_MODE.equals(type)) {
            return 16;
        } else if (CameraParameterType.PRIVACY.equals(type)) {
            return 17;
        }
        return -1;
    }

    public static <T> T getSupportedParameter(long nativeId, CameraSupportParameters.Key<T> key) {
        if (nativeId == 0L) return null;
        String result = nativeGetSupportedParameters(nativeId, ketTypeToNativeId(key.type));
        if (key.type.equals(CameraParameterType.PREVIEW_SIZE)) {
            if (TextUtils.isEmpty(result)) {
                return null;
            }
            try {
                JSONArray jsonArray = new JSONArray(result);
                if (jsonArray.length() <= 0) {
                    return null;
                }
                List<CameraSize> lists = new ArrayList<>();
                for (int i = 0; i < jsonArray.length(); i++) {
                    JSONObject jsonObject = jsonArray.optJSONObject(i);
                    int width = jsonObject.optInt("width");
                    int height = jsonObject.optInt("height");
                    int format = jsonObject.optInt("format");
                    lists.add(new CameraSize(width, height, CameraPreviewFormat.valueToFormatMode(format)));
                }
                return (T) lists.toArray(new CameraSize[0]);
            } catch (Exception e) {
                return null;
            }
        }

        if (boolean.class.isAssignableFrom(key.mClass) || Boolean.class.isAssignableFrom(key.mClass)) {
            int state = 0;
            if (!TextUtils.isEmpty(result)) {
                state = Integer.parseInt(result);
            }
            return (T) Boolean.valueOf(state == 1);
        } else if (IntRange.class.isAssignableFrom(key.mClass)) {
            if (TextUtils.isEmpty(result) || !result.contains(":")) {
                return null;
            }
            String[] split = result.split(":");
            if (split.length != 2) {
                return null;
            }
            return (T) new IntRange(Integer.parseInt(split[0]), Integer.parseInt(split[1]));
        }
        return null;
    }

    public static native boolean nativeStartRecord(long nativeId, String fileName);

    public static native boolean nativeStopRecord(long nativeId);

    public static native void nativeSetRecordFormat(long nativeId, int format);

    public static native void nativeSetParentPath(long nativeId, String parentPath);

    public static native boolean nativeSetDefaultParentPath(String parentPath);

    public static native String nativeGetRecordPath(long nativeId);
}

File: libusbcamera/src/main/java/com/rain/uvc/utils/CameraNativeUtils.kt
package com.rain.uvc.utils

import android.view.Surface
import com.rain.uvc.listener.IButtonListener
import com.rain.uvc.listener.IFrameListener
import com.rain.uvc.mode.CameraSize
import com.rain.uvc.mode.CameraSupportSize
import com.rain.uvc.parameters.CameraParameterType
import com.rain.uvc.parameters.CameraPreviewFormat
import com.rain.uvc.parameters.Parameters
import com.rain.uvc.parameters.SupportParameters
import org.json.JSONArray

/**
 * @author yuan
 * @createTime: 2026/2/2
 * @des usb相机调用native层接口
 */
object CameraNativeUtils {
	init {
		System.loadLibrary("usb100")
		System.loadLibrary("uvc")
		System.loadLibrary("uvcCamera")
	}
	
	/**
	 * 初始化debug模式
	 */
	@JvmStatic
	external fun debuggable(status: Int): Boolean
	
	/**
	 * 读取可使用的v4l2的列表信息
	 *
	 * @return 返回列表地址
	 */
	@JvmStatic
	external fun nativeLoadV4L2Devices(): Array<String>?
	
	/**
	 * 根据对应的FileDescriptor连接指定设备
	 *
	 * @param fd 文件描述符
	 * @return 是否成功
	 */
	@JvmStatic
	external fun nativeOpen(fd: Int, busNum: Int, devAddress: Int): Long
	
	/**
	 * 根据对应的FileDescriptor连接指定设备
	 *
	 * @param videoPath 对应打开的路径
	 * @return 是否成功
	 */
	@JvmStatic
	external fun nativeOpenVideo(videoPath: String?): Long
	
	/**
	 * 断开连接设备
	 *
	 * @param nativeId 对应的内存地址值
	 * @return 是否成功
	 */
	@JvmStatic
	external fun nativeClose(nativeId: Long): Boolean
	
	/**
	 * 开启预览
	 *
	 * @param nativeId 对应的内存地址值
	 * @return 是否成功
	 */
	@JvmStatic
	external fun nativeStartPreview(nativeId: Long): Boolean
	
	/**
	 * 关闭预览
	 *
	 * @param nativeId 对应的内存地址值
	 * @return 是否成功
	 */
	@JvmStatic
	external fun nativeStopPreview(nativeId: Long): Boolean
	
	/**
	 * 设置预览控件
	 *
	 * @param nativeId 对应的内存地址值
	 * @param surface  对应预览的surface
	 * @return 是否成功
	 */
	@JvmStatic
	external fun nativeSetDisplaySurface(nativeId: Long, surface: Surface?): Boolean
	
	/**
	 * 设置预览的分辨率信息
	 *
	 * @param nativeId 对应的内存地址值
	 * @param width    宽
	 * @param height   高
	 * @return 是否成功
	 */
	@JvmStatic
	external fun nativeSetPreviewSize(nativeId: Long, width: Int, height: Int, format: Int): Boolean
	
	/**
	 * 设置预览监听，返回的数据格式固定为RGB格式
	 *
	 * @param nativeId 设置的对应id
	 * @param listener 监听器
	 */
	@JvmStatic
	external fun setPreviewListener(nativeId: Long, listener: IFrameListener?, mode: Int): Boolean
	
	/**
	 * 设置按钮监听
	 * @param nativeId 设置的对应id
	 * @param listener 监听器
	 */
	@JvmStatic
	external fun setButtonListener(nativeId: Long, listener: IButtonListener?): Boolean
	
	/**
	 * 获取支持的参数信息
	 */
	@JvmStatic
	external fun nativeGetSupportedParameters(nativeId: Long, type: Int): String?
	
	/**
	 * 获取当前参数信息
	 */
	@JvmStatic
	external fun nativeGetParameterValue(nativeId: Long, type: Int): String?
	
	/**
	 * 设置当前参数信息
	 */
	@JvmStatic
	external fun nativeSetParameterValue(nativeId: Long, type: Int, value: Int): Boolean
	
	/**
	 * 执行拍照
	 */
	@JvmStatic
	external fun nativeTakePicture(nativeId: Long): ByteArray?
	
	/**
	 * 获取对应的参数信息
	 *
	 * @param nativeId 对应的内存地址值
	 * @param key      对应支持的参数key
	 * @return 返回值，如果不支持或者不存在，则返回null
	 */
	@JvmStatic
	fun <T> getParameter(nativeId: Long, key: Parameters.Key<T>): T? {
		if (nativeId == 0L) return null
		// 获取到native层返回的数据
		val returnValue = nativeGetParameterValue(nativeId, key.type.strToKeyId())
		if (returnValue.isNullOrBlank()) return null
		return when (key.mClass) {
			CameraSize::class.java -> {
				val sizeArray = returnValue.split(":")
				if (sizeArray.size < 2) return null
				CameraSize(
					sizeArray[0].toInt(), sizeArray[1].toInt()
				)
			}
			String::class.java -> returnValue
			Int::class.java, Int::class.javaPrimitiveType -> returnValue.toIntOrNull()
			Boolean::class.java, Boolean::class.javaPrimitiveType -> returnValue.toIntOrNull() == 1
			Double::class.java, Double::class.javaPrimitiveType -> returnValue.toDoubleOrNull()
			Float::class.java, Float::class.javaPrimitiveType -> returnValue.toFloatOrNull()
			else -> null
		} as? T
	}
	
	/**
	 * 设置对应的参数信息
	 *
	 * @param nativeId 对应的内存地址值
	 * @param key      对应支持的参数key
	 * @param value    对应的值
	 * @return 是否支持
	 */
	fun <T> setParameter(nativeId: Long, key: Parameters.Key<T>, value: T): Boolean {
		if (nativeId == 0L) return false
		val keyId = key.type.strToKeyId()
		if (keyId == -1) return false
		return when (key.mClass) {
			Int::class.java, Int::class.javaPrimitiveType -> nativeSetParameterValue(nativeId, keyId, value as Int)
			Boolean::class.java, Boolean::class.javaPrimitiveType -> nativeSetParameterValue(nativeId, keyId, if (value as Boolean) 1 else 0)
			else -> false
		}
	}
	
	/**
	 * 获取对应支持的参数列表信息
	 * @param nativeId 对应的内存地址值
	 * @param key      对应支持的参数key
	 * @return 对应支持的列表
	 */
	fun <T> getSupportedParameter(nativeId: Long, key: SupportParameters.Key<T>): T? {
		if (nativeId == 0L) return null
		val keyId = key.type.strToKeyId()
		if (keyId == -1) return null
		val returnValue: String? = nativeGetSupportedParameters(nativeId, keyId)
		if (returnValue.isNullOrBlank()) return null
		return when (key.mClass) {
			Array<CameraSupportSize>::class.java -> runCatching {
				val groupArray = JSONArray(returnValue)
				if (groupArray.length() <= 0) return@runCatching null
				return@runCatching Array(groupArray.length()) {
					val jsonObject = groupArray.optJSONObject(it)
					val format = jsonObject.optInt("format")
					val sizes = jsonObject.optJSONArray("sizes")
					val sizeList = if (sizes != null && sizes.length() > 0) {
						List(sizes.length()) { childIndex ->
							val childJsonObj = sizes.getJSONObject(childIndex)
							val width = childJsonObj.getInt("width")
							val height = childJsonObj.getInt("height")
							CameraSize(width, height)
						}
					} else listOf()
					return@Array CameraSupportSize(
						CameraPreviewFormat.valueToFormatMode(format), sizeList
					)
				}
			}.getOrNull()
			Boolean::class.java, Boolean::class.javaPrimitiveType -> returnValue.toInt() == 1
			IntRange::class.java -> {
				if (returnValue.contains(":")) {
					val split = returnValue.split(":")
					if (split.size == 2) IntRange(split[0].toInt(), split[1].toInt()) else null
				} else null
			}
			else -> null
		} as? T
	}
}

/**
 * 将当前的key转为对应native层需要的id
 */
private fun String.strToKeyId(): Int {
	return when (this) {
		CameraParameterType.PREVIEW_SIZE -> 0        // 预览分辨率
		CameraParameterType.PREVIEW_ORIENTATION -> 1         // 预览方向
		CameraParameterType.AUTO_EXPOSURE -> 2       // 自动曝光
		CameraParameterType.EXPOSURE -> 3            // 曝光度
		CameraParameterType.BRIGHTNESS -> 4          // 亮度
		CameraParameterType.CONTRAST -> 5            // 对比度
		CameraParameterType.GAIN -> 6                // 增益值
		CameraParameterType.SATURATION -> 7          // 饱和度
		CameraParameterType.ZOOM -> 8                // 缩放
		CameraParameterType.AUTO_FOCUS -> 9          // 自动对焦
		CameraParameterType.FOCUS -> 10              // 焦距
		CameraParameterType.IRIS -> 11               // 光圈
		CameraParameterType.AUTO_HUE -> 12           // 自动变化色调
		CameraParameterType.HUE -> 13                // 色调
		CameraParameterType.AUTO_WHITE_BALANCE -> 14 // 自动白平衡
		CameraParameterType.WHITE_BALANCE -> 15      // 白平衡
		CameraParameterType.SCENE_MODE -> 16         // 场景模式
		CameraParameterType.PRIVACY -> 17            // 隐私模式
		CameraParameterType.JPEG_MIRROR -> 18             // 镜像处理
		else -> -1
	}
}